### PR TITLE
Add caption support to the Gallery block in <amp-carousel> and fix displaying Gallery block as carousel in WP 5.3

### LINKED ISSUE #2855: Add image captions to the gallery carousel
Originally reported by @rsmith4321 on https://wordpress.org/support/topic/amp-carousel-with-captions/:

> Is it possible to add captions the images in the amp gallery carousel? I know it is possible here https://amp.dev/documentation/examples/multimedia-animations/image_galleries_with_amp-carousel/ but it looks like it requires adding a caption div to each slide. Maybe this could be a future feature, thanks.

Should the captions be shown in the carousel slides by default or should there be a toggle to enable them?

### PR DESCRIPTION
**Pull request by kienstra**
### Update: this has changed a lot, please see [more recent comments](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-546171299)

For images in a gallery that have a caption, this adds caption support when 'Display as carousel' is selected:
captions-in-carousel

# Before

before-caption

# After

caption-appears-now

Closes #2855

(Please do `npm run build:css` for the CSS here)

### CODE REVIEW

**Comment by kienstra** [comment]
There are probably more styling considerations with this. Sometimes the caption overlaps the `<amp-img>`. 

Simply placing the caption on top of the `<amp-img>`, with a scrim behind it, could be an option. But the scrim doesn't look good with certain themes.

**Comment by kienstra** [comment]
**Couldn't Get Styling To Work With Core Themes**
*Maybe a CSS specialist could help*

I couldn't get the captions to look good on all of the Core themes, and having them look good on *all* themes would probably be even harder.

Maybe a CSS specialist could help here.

Also, I tried placing the caption above the `<amp-img>` and [adding a scrim](https://github.com/ampproject/amp-wp/pull/3285/commits/0b0b2fd714987b465c066e69dbb3af7ad6f318f1) with a black gradient. But sometimes it was too wide:

scrim-wide

Here are the displays with the current state of the PR (without the scrim):

# Twenty Fifteen
above-image

# Twenty Sixteen

twenty-sixteen

# Twenty Seventeen
twenty-seventeen-gallery

# Twenty Nineteen

2019-gallery

Steps To Reproduce
1. Create a new post
2. Add a gallery block
3. Upload 3 [600 x 400 images](https://picsum.photos/600/400)
4. Add captions to some of the images
5. Select 'Display as carousel'
6. Save the post, and go to the front-end

Of course, galleries could have any combination of image sizes, but the screenshots above all use [600 x 400 images](https://picsum.photos/600/400) to simplify this.

**Comment by kienstra** [comment]
# Before 
The plugin output the `<amp-carousel>` like:

```php
<amp-carousel type="slides" ...>
        <amp-img ...>
</amp-carousel>
```

# After (with captions)

With this PR, the plugin outputs something like:

```php
<amp-carousel type="slides" ...>
        <div class="slide">
                <amp-img ...>
                <div class="amp-wp-gallery-caption">
                        This is a caption
                </div>
        </div>
</amp-carousel>
```

(Though the AMP runtime adds a lot of tags, like `<div class="i-amphtml-slide-item i-amphtml-slide-item-show">`)

It looks like [amp-carousels with captions](https://amp.dev/documentation/examples/multimedia-animations/image_galleries_with_amp-carousel/?format=websites) have to have something like a `<div class="slide">` wrapper for each slide.

But on the Core themes like Twenty Fifteen and Twenty Seventeen, I [couldn't find a way](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-533706512) to style the caption so it appears in the same place, or at least looks good.

And ensuring that captions appear good on *any* theme would probably be much harder.

I'm not sure if there's a solution for this 😄

**Comment by kienstra** [comment]
Also, please do `npm run build:css` to compile the CSS locally.

**Comment by austintude** [comment]
I've been unable, time wise, to look into this more deeply but the prospect of not only adding but modifying these is of great importance to several of my clients.
In an ideal world, I'd be able to change: font-size, margin, padding, alignment, and other layout typical styling options availble with most text based content. 
While the multi-level-wrapping-text is clickable and expands up after being clicked, it is so low on the screen that this function is NOT obvious to the end user/site-visitor. /
Thank you for working on this. Keep up the good work.
Daniel

**Comment by westonruter** [comment]
@kienstra I suppose this will then need UX support.

**Comment by kienstra** [comment]
Hi @westonruter,
Yeah, that makes sense.

If it's alright, we've asked Bart Makos to work on this, as he's much better with CSS than me.

Though I'm [not sure there is a good solution](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-533706512) for captions, where they'll look good on any theme.

**Comment by austintude** [comment]
This is starting to sound fun. I love the word CSS!!!

**Comment by kienstra** [comment]
Haha, I wish I was great at CSS 😄

**Comment by westonruter** [comment]
This may have some inspiration: https://github.com/Automattic/newspack-blocks/pull/143

**Comment by kienstra** [comment]
Nice, thanks for posting that! It could help.

**Comment by kienstra** [comment]
Thanks to [Weston's suggestion](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-539672495), captions look good so far. Though there's still some work left.

gallery-here

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php [hunk 1]
```diff
@@ -140,8 +140,41 @@ public function sanitize() {
 					'layout' => 'responsive',
 				]
 			);
+
 			foreach ( $images as $image ) {
-				$amp_carousel->appendChild( $image );
+				$div = AMP_DOM_Utils::create_node(
```
Comment (kienstra) [inline includes/sanitizers/class-amp-gallery-block-sanitizer.php]: This used to simply append the image to the `<amp-carousel>`. But now it creates a `<div class="slide">` wrapper for the image, and possibly the caption.

It looks like that's needed in order to display a caption.

**Comment by kienstra** [comment]
**Gallery Blocks With <amp-carousel> Sometimes Look Different**
*Especially square images*

Some Gallery blocks look different with this PR. But it looks like they better display the images at the `width` and `height` that the `AMP_Img_Sanitizer` produces.

# Before

before-g

# After

after-g

The `<amp-img>` above looks like a rectangle, as you'd expect from its `width=“640” height=“400”`.

This is from the new [object-fit: cover](https://github.com/ampproject/amp-wp/pull/3285/files#diff-3f3a86f42d679f81a48116543bcb3b38R154) for the `<amp-img>`, which prevents the caption from overflowing the image:

scrim-wide

## Steps to reproduce
1. Create a new post
2. Add a gallery block
3. Upload few [800 x 800 images](https://picsum.photos/800/800)
4. Select 'Display as carousel'
5. Save the post, and go to the front-end of the AMP URL

**Comment by kienstra** [comment]
## Steps to test

1. Create a new post with a Gallery block
2. Upload few [rectangular](https://picsum.photos/600/400) and [square images](https://picsum.photos/800/800) to it
4. Add captions to some, including very long captions
3. Select 'Display as carousel'
4. Save the post, and go to the front-end of the AMP URL
5. Expected: the captions always appear over the image, no matter how long they are:

captions-gallery

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php [hunk 2]
```diff
@@ -140,8 +140,43 @@ public function sanitize() {
...
-				$amp_carousel->appendChild( $image );
+				$slide = AMP_DOM_Utils::create_node(
+					$this->dom,
+					'div',
+					[ 'class' => 'slide' ]
+				);
+
+				// Ensure the image fills the entire <amp-carousel>, so the possible caption looks right.
+				if ( 'amp-img' === $image->tagName ) {
+					$image->setAttribute( 'layout', 'fill' );
```
Comment (kienstra) [inline includes/sanitizers/class-amp-gallery-block-sanitizer.php]: `layout="fill"` appears a lot in the documentation for [Image Galleries with amp-carousel](https://amp.dev/documentation/examples/multimedia-animations/image_galleries_with_amp-carousel/?format=websites)

**Comment by kienstra** [comment]
**Request For Review**

Hi @westonruter,
Good to see you earlier today.

Could you please review this? Your [suggestion](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-539672495) to look at Newspack really helped. 

There is a [possible backwards-compatibility issue](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-546168357). I think it's probably an improvement, but people might not like that it's different.

Also, I could style the caption wrapper a little differently. Like you mentioned, maybe this [could use UX help](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-537267092):

caption-wrapper

In my testing, this looks good with the Core themes from Twenty Fifteen to Twenty Twenty.

You probably know this better than me, but this will require doing `npm run build:css` for the CSS here.

Thanks, Weston!

**Comment by westonruter** [comment]
@kienstra I'm trying to test this but I'm not seeing the carousel apply. I have configured the Gallery block like so:

image

It's not getting converted into a carousel:

image

It seems this is due to [this change](https://make.wordpress.org/core/2019/09/27/block-editor-theme-related-updates-in-wordpress-5-3/):

image

**Comment by kienstra** [comment]
>I took an initial stab at adding compatibility, but some tests are broken now and need to be fixed. The changes need to be in WordPress 5.2 without Gutenberg active as well, to ensure the carousel is displaying as expected.

Hi Weston, thanks for making those commits. Ah, I should have tested this with WordPress 5.3

I'll work on the tests.

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php [hunk 3]
```diff
@@ -73,23 +73,40 @@ class AMP_Gallery_Block_Sanitizer extends AMP_Base_Sanitizer {
...
-			return;
-		}
+		$xpath = new DOMXPath( $this->dom );
+		$expr  = sprintf(
+			'//ul[ %s ]',
+			implode(
+				' or ',
+				[
+					'( parent::figure[ contains( @class, "wp-block-gallery" ) and @data-amp-carousel = "true" ] )',
+					'( contains( @class, "wp-block-gallery" ) and @data-amp-carousel = "true" )',
```
Comment (claudiulodro) [inline includes/sanitizers/class-amp-gallery-block-sanitizer.php]: I believe you've lost some functionality with this change. Previously you could have the carousel "off" and lightbox "on" for the gallery, but this sanitizer will now ignore that setting combination.

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php [hunk 4]
```diff
@@ -73,23 +73,40 @@ class AMP_Gallery_Block_Sanitizer extends AMP_Base_Sanitizer {
...
-			}
+		foreach ( $nodes as $node ) {
+			/**
+			 * Element
+			 *
+			 * @var DOMElement $node
+			 */
 
 			$attributes      = AMP_DOM_Utils::get_node_attributes_as_assoc_array( $node );
 			$is_amp_lightbox = isset( $attributes['data-amp-lightbox'] ) && true === filter_var( $attributes['data-amp-lightbox'], FILTER_VALIDATE_BOOLEAN );
```
Comment (claudiulodro) [inline includes/sanitizers/class-amp-gallery-block-sanitizer.php]: Like `$is_amp_carousel`, `$is_amp_lightbox` also needs to check the `parentNode` to see if `data-amp-lightbox` is set (for Gutenberg 6.5+ compatibility). The simplest way of handling this situation for both `$is_amp_carousel` and `$is_amp_lightbox` might be something like this:
```
$attributes      = array_merge( AMP_DOM_Utils::get_node_attributes_as_assoc_array( $node ), AMP_DOM_Utils::get_node_attributes_as_assoc_array( $node->parentNode ) );
```

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php [hunk 5]
```diff
@@ -73,23 +73,40 @@ class AMP_Gallery_Block_Sanitizer extends AMP_Base_Sanitizer {
...
-		if ( 0 === $num_nodes ) {
-			return;
-		}
+		$xpath = new DOMXPath( $this->dom );
+		$expr  = sprintf(
+			'//ul[ %s ]',
+			implode(
+				' or ',
+				[
+					'( parent::figure[ contains( @class, "wp-block-gallery" ) and @data-amp-carousel = "true" ] )',
```
Comment (westonruter) [inline includes/sanitizers/class-amp-gallery-block-sanitizer.php]: The `contains( @class… )` logic here and the line below should be refactored to make use of `normalize-space()` for example:

```
concat( ' ', normalize-space( @class ), ' ' ), ' wp-block-gallery ' )
```

Comment (kienstra) [reply includes/sanitizers/class-amp-gallery-block-sanitizer.php]: Thanks, @westonruter. Is https://github.com/ampproject/amp-wp/pull/3285/commits/9af22c82b6f7abecc4be3d2c9f4e0668ac72fd39 like what you had in mind?

File: includes/sanitizers/class-amp-gallery-block-sanitizer.php (hunk 4 above)
Comment (kienstra) [reply includes/sanitizers/class-amp-gallery-block-sanitizer.php]: Hi @claudiulodro, thanks for pointing that out. I ended up doing something similar to that.

**Comment by kienstra** [comment]
Hi @westonruter,
Thanks for your commits, and for testing this. 

In my testing, it's looking good with the following combinations:

### Environment 
1. With Gutenberg active (`6.7.0`, latest stable) 
2. Without Gutenberg, running WP Core `5.2.4`

### Gallery Block Settings
1. Only lightbox 
2. Only carousel
3. Lightbox and carousel
4. With and without captions

Though it looks like all Gallery blocks appear with a lightbox, whether it's selected or not. This issue also seems to appear on the `develop` branch and the last stable release (`1.3.0` tag):

light-box-gallery

**Comment by kienstra** [comment]
>I noticed a couple small bugs in the implementation. In general, though, this works great with and without Gutenberg 6.5+!

Hi @claudiulodro,
Thanks a lot for testing this. If you have time, it'd be great if you could test this again. This should now work when the carousel is off and the lightbox is on.

**Comment by kienstra** [comment]
One thing I missed is that this doesn't add captions to gallery shortcodes with 'Display as carousel' selected, like:
```
[gallery ids="4572, 3619"]
```

**Comment by westonruter** [comment]
> One thing I missed is that this doesn't add captions to gallery shortcodes with 'Display as carousel' selected

@kienstra So the only outstanding thing is classic `gallery` shortcode instances will continue to be rendered without captions? If that's the case, then I think we can merge this and then follow-up with another issue/PR to do the rest.

**Comment by kienstra** [comment]
Hi @westonruter,
>@kienstra So the only outstanding thing is classic gallery shortcode instances will continue to be rendered without captions?

That's right, this PR doesn't affect the Gallery shortcode (or the Classic editor galleries, which use that shortcode).

**Comment by westonruter** [comment]
OK great, so `[gallery]` shortcode use will appear no different for users, while at the same time this change fixes the Gutenberg gallery block integration to ensure that the Carousel setting actually has an effect:

<table><thead><tr><th>Non-AMP<th>AMP Before<th>AMP After</tr>
<tbody><tr valign=top><td>Screen Shot 2019-10-28 at 14 08 12<td>Screen Shot 2019-10-28 at 14 08 24<td>Screen Shot 2019-10-28 at 14 08 48</tr></tbody></table>

**Comment by kienstra** [comment]
>@kienstra So the only outstanding thing is classic gallery shortcode instances will continue to be rendered without captions? If that's the case, then I think we can merge this and then follow-up with another issue/PR to do the rest.

Hi @westonruter,
Thanks a lot for reviewing this. Would it be alright if I opened an issue for `[gallery]` shortcode captions, then worked on a PR for that? 

The [support topic](https://wordpress.org/support/topic/amp-carousel-with-captions/) that prompted this PR's issue (#2855) looks to use a Gallery shortcode.

Thanks, Weston.

**Comment by austintude** [comment]
Great stuff y'all!!

**Comment by westonruter** [comment]
@kienstra por supuesto. 👍

**Comment by kienstra** [comment]
Gracias!

**Pull request merged**